### PR TITLE
KAFKA-16741: Add share group classes for Heartbeat API (1/N) (KIP-932)

### PR DESCRIPTION
**Pull request by apoorvmittal10**
Defined share group, member and sinmple assignor classes with API definition for Share Group Heartbeat and Describe API.

The `ShareGroup` and `ShareGroupMember` extends the common `ModernGroup` and `ModernGroupMember` respectively.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: group-coordinator/src/main/resources/common/message/ShareGroupMemberMetadataKey.json [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// KIP-932 is in development. This schema is subject to non-backwards-compatible changes.
+{
+  "type": "data",
+  "name": "ShareGroupMemberMetadataKey",
+  "validVersions": "10",
```
Comment (apoorvmittal10) [inline group-coordinator/src/main/resources/common/message/ShareGroupMemberMetadataKey.json:20]: Though we have keys defined till `8` and shortly there will be another key defined at version `9`, as part of KIP-932 itself. Hence as per KIP, kept this key at version `10`.

Comment (AndrewJSchofield) [reply group-coordinator/src/main/resources/common/message/ShareGroupMemberMetadataKey.json:20]: I would expect `ShareGroupMetadataKey` and `ShareGroupMetadataValue` in the KIP to be defined at the same time. I'd prefer not to leave gaps in the versions, so if we are introducing a subset of the KIP's record schemas, I suggest renumbering them in the KIP so we get a continuous sequence once the PR is merged.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/share/ShareGroup.java [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+    }
+
+    /**
+     * Populates the list of records with tombstone(s) for deleting the group.
+     *
+     * @param records The list of records.
+     */
+    @Override
+    public void createGroupTombstoneRecords(List<CoordinatorRecord> records) {
+        records.add(CoordinatorRecordHelpers.newGroupEpochTombstoneRecord(groupId()));
```
Comment (AndrewJSchofield) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/share/ShareGroup.java]: This should actually be created a `ShareGroupMetadata` tombstone.

Comment (AndrewJSchofield) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/share/ShareGroup.java]: And also, the GroupMetadataManager should be writing `ShareGroupMetadata` records in `shareGroupHeartbeat`, when we get to that.

**Review by AndrewJSchofield — CHANGES_REQUESTED** [review]
Thanks for the PR. A few minor comments.

File: group-coordinator/src/main/resources/common/message/ShareGroupMemberMetadataKey.json (hunk 1 above)
Comment (apoorvmittal10) [reply group-coordinator/src/main/resources/common/message/ShareGroupMemberMetadataKey.json:20]: Done.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/share/ShareGroup.java (hunk 2 above)
Comment (apoorvmittal10) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/share/ShareGroup.java]: You are right, corrected it.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/CoordinatorRecordHelpers.java [hunk 3]
```diff
@@ -209,14 +211,40 @@ public static CoordinatorRecord newGroupEpochRecord(
...
+                    (short) 3
+                ),
+                null // Tombstone.
+            );
+        } else if (groupType == GroupType.SHARE) {
+            return new CoordinatorRecord(
+                new ApiMessageAndVersion(
+                    new ShareGroupMetadataKey()
+                        .setGroupId(groupId),
+                    (short) 3
```
Comment (AndrewJSchofield) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/CoordinatorRecordHelpers.java]: This should be 11.

**Review by AndrewJSchofield — CHANGES_REQUESTED** [review]

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/CoordinatorRecordHelpers.java (hunk 3 above)
Comment (apoorvmittal10) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/CoordinatorRecordHelpers.java]: My bad, corrected and added tests as well.

**Review by AndrewJSchofield — APPROVED** [review]

**Review by omkreddy — APPROVED** [review]
@apoorvmittal10 Thanks for the PR. LGTM

**Comment by apoorvmittal10** [comment]
Build passed with unrelated tests failure.

**Comment by C0urante** [comment]
@apoorvmittal10 @omkreddy there are compilation failures on trunk in the `SimpleAssignorTest` suite caused by a constructor change to the `MemberSubscriptionAndAssignmentImpl ` class introduced in https://github.com/apache/kafka/pull/16504. I know that this isn't the fault of anyone involved in this PR since it was merged before the one that led to these compilation failures, but if either of you has a moment, would it be possible to patch the test suite to fix the build on trunk?

**Comment by C0urante** [comment]
Ah, spoke too soon--looks like a fix has been published: https://github.com/apache/kafka/pull/16526

**Pull request merged**